src/client/Microsoft.Identity.Client/ApiConfig/AbstractAcquireTokenParameterBuilder.cs

```diff
@@ -1,379 +1,432 @@
+        /// </summary>
+        /// <param name="tenantId">The tenant ID, which can be either in GUID format or a domain name. Also known as the Directory ID.</param>
+        /// <returns>The builder to chain the .With methods.</returns>
+        /// <exception cref="ArgumentNullException">Thrown if tenantId is null or an empty string</exception>
+        /// <exception cref="MsalClientException">Thrown if the application was configured with an authority that is not AAD specific (e.g. ADFS or B2C).</exception>
+        /// <remarks>
+        /// The tenant should be more restrictive than the one configured at the application level, e.g. don't use "common".
+        /// Does not affect authority validation, which is specified at the application level.</remarks>
+        public T WithTenantId(string tenantId)
+        {
+            if (string.IsNullOrEmpty(tenantId))
+            {
+                throw new ArgumentNullException(nameof(tenantId));
+            }
+
+            if (!ServiceBundle.Config.Authority.AuthorityInfo.IsTenantOverrideSupported)
+            {
+                throw new MsalClientException(
+                    MsalError.TenantOverrideNonAad,
+                    MsalErrorMessage.TenantOverrideNonAad);
+            }
+
+            AadAuthority aadAuthority = (AadAuthority)ServiceBundle.Config.Authority;
+            string tenantedAuthority = aadAuthority.GetTenantedAuthority(tenantId, true);
+            var newAuthorityInfo = AuthorityInfo.FromAadAuthority(
+                tenantedAuthority,
+                ServiceBundle.Config.Authority.AuthorityInfo.ValidateAuthority);
+
+            CommonParameters.AuthorityOverride = newAuthorityInfo;
+
+            return (T)this;
+        }
+
+        /// <summary>
+        /// Overrides the tenant ID specified in the authority at the application level. This operation preserves the authority host (environment).
+        /// 
+        /// If an authority was not specified at the application level, the default used is https://login.microsoftonline.com/common.
+        /// </summary>
+        /// <param name="authority">Uri for the authority that includes the TenantId</param>
+        /// <returns>The builder to chain the .With methods.</returns>
+        /// <exception cref="ArgumentNullException">Thrown if <paramref name="authority"/> is null or an empty string</exception>
+        /// <exception cref="MsalClientException">Thrown if the application was configured with an authority that is not AAD specific (e.g. ADFS or B2C).</exception>
+        /// <remarks>
+        /// The tenant should be more restrictive than the one configured at the application level, e.g. don't use "common".
+        /// Does not affect authority validation, which is specified at the application level.</remarks>
+        public T WithTenantIdFromAuthority(string authority)

```

Comment:
nit: should we stop giving the option to use strings for URIs? I remember reading somewhere about this or seeing some C# analyzer suggesting this.

Reply:
Hmm, +1? We can always add it, if vehemently requested.

Reply:
@bgavrilMS  @pmaytak  
I added it mainly for consistency with the other API's, but I don't mind removing it

Reply:
It's rule CA2234. I've removed it now